paywall/src/data-iframe/blockchainHandler/keyStatus.js

```diff
@@ -0,0 +1,63 @@
+import { getAccount } from './account'
+
+export function validKey(key) {
+  return key.expiration > new Date().getTime() / 1000
+}
+
+export function getKeyStatus(key, requiredConfirmations) {
+  if (!key.transactions || !key.transactions.length) {
+    return validKey(key) ? 'valid' : 'none'
+  }
+  const lastTransaction = key.transactions[0]
+  switch (lastTransaction.status) {
+    case 'mined':
+      if (lastTransaction.confirmations < requiredConfirmations) {
+        return 'confirming'
+      }
+      if (key.expiration < new Date().getTime() / 1000) {
+        return 'expired'
+      }
+      return 'valid'
+    default:
+      return lastTransaction.status || 'none'
+  }
+}
+
+export function linkTransactionsToKeys({
+  keys,
+  transactions,
+  locks,
+  requiredConfirmations,
+}) {
+  const account = getAccount()
+  const newKeys = {}
+
+  const transactionsByLock = locks.reduce((indexedTransactions, lock) => {
+    // get the key purchase transactions sorted in reverse chronological order
+    const keyPurchaseTransactions = Object.values(transactions)
+      .filter(transaction => {
+        return transaction.from === account && transaction.to === lock
+      })
+      .sort((a, b) => (a.blockNumber > b.blockNumber ? -1 : 1))
```

Comment:
Pending transactions do not have a blockNumber... I want to make sure we handle this case correctly.

Reply:
In `web3Service._getSubmittedTransaction`, we do:

```js

    return this.emit('transaction.updated', transactionHash, {
      status: 'submitted',
      confirmations: 0,
      blockNumber: Number.MAX_SAFE_INTEGER, // Asign the largest block number for sorting purposes
    })
```

which covers this situation. pending transactions inherit this huge block number, which finally gets overwritten with the correct one when it gets mined.

Reply:
Ha! thanks for looking this up.